cpp/src/io/text/multibyte_split.cu

```diff
@@ -0,0 +1,543 @@
+// multibyte_split works by splitting up inputs in to 32 inputs (bytes) per thread, and transforming
+// them in to data structures called "multistates". these multistates are created by searching a
+// trie, but instead of a tradition trie where the search begins at a single node at the beginning,
+// we allow our search to begin anywhere within the trie tree. The position within the trie tree is
+// stored as a "partial match path", which indicates "we can get from here to there by a set of
+// specific transitions". By scanning together multistates, we effectively know "we can get here
+// from the beginning by following the inputs". By doing this, each thread knows exactly what state
+// it begins in. From there, each thread can then take deterministic action. In this case, the
+// deterministic action is counting and outputting delimiter offsets when a delimiter is found.
+
+__global__ void multibyte_split_init_kernel(cudf::size_type base_tile_idx,
+                                            cudf::size_type num_tiles,
+                                            scan_tile_state_view<multistate> tile_multistates,
+                                            scan_tile_state_view<uint32_t> tile_output_offsets,
+                                            scan_tile_status status = scan_tile_status::invalid)
+{
+  tile_multistates.initialize_status(base_tile_idx, num_tiles, status);
+  tile_output_offsets.initialize_status(base_tile_idx, num_tiles, status);
+}
+
+__global__ void multibyte_split_kernel(cudf::size_type base_tile_idx,
+                                       cudf::size_type num_tiles,
+                                       scan_tile_state_view<multistate> tile_multistates,
+                                       scan_tile_state_view<uint32_t> tile_output_offsets,
+                                       cudf::io::text::trie_device_view trie,
+                                       cudf::device_span<char const> data,
+                                       cudf::device_span<int32_t> string_offsets,
+                                       cudf::device_span<char> data_out)
+{
+  typedef cub::BlockScan<uint32_t, THREADS_PER_TILE> OffsetScan;
+  typedef scan_tile_state_callback<uint32_t> OffsetScanCallback;
+
+  __shared__ union {
+    typename PatternScan::TempStorage pattern_scan;
+    struct {
+      typename OffsetScan::TempStorage offset_scan;
+      typename OffsetScanCallback::TempStorage offset_scan_callback;
+    };
+  } temp_storage;
+
+  int32_t const tile_idx   = base_tile_idx + blockIdx.x;
+  int32_t const thread_idx = blockIdx.x * blockDim.x + threadIdx.x;
+  int32_t const data_begin = thread_idx * ITEMS_PER_THREAD;
+  int32_t const num_valid  = data.size() - data_begin;
+  int32_t const char_begin = base_tile_idx * ITEMS_PER_TILE;
+
+  // STEP 1: Load inputs
+
+  char thread_data[ITEMS_PER_THREAD];
+
+  for (int32_t i = 0; i < ITEMS_PER_THREAD and i < num_valid; i++) {  //
+    thread_data[i] = data[data_begin + i];
+  }
```

Comment:
Use block load? This is doing non-coalesced 1B reads.

Reply:
Never got around to it, but I can do it now easy enough.

Reply:
Updated to use `cub::BlockLoad` with `cub::BLOCK_LOAD_VECTORIZE`.